nucypher/blockchain/eth/sol/source/contracts/Adjudicator.sol

```diff
@@ -72,77 +72,77 @@ contract Adjudicator is Upgradeable {
     }
 
     /**
-    * @notice Submit proof that staker created wrong CFrag
+    * @notice Submit proof that a worker created wrong CFrag
     * @param _capsuleBytes Serialized capsule
     * @param _cFragBytes Serialized CFrag
-    * @param _cFragSignature Signature of CFrag by staker
+    * @param _cFragSignature Signature of CFrag by worker
     * @param _taskSignature Signature of task specification by Bob
-    * @param _requesterPublicKey Requester's public key that was used to sign Capsule
-    * @param _stakerPublicKey Staker's public key that was used to sign Capsule and CFrag
-    * @param _stakerPublicKeySignature Signature of public key by staker's eth-key
-    * @param _preComputedData Pre computed data for CFrag correctness verification
+    * @param _requesterPublicKey Bob's signing public key, also known as "stamp"
+    * @param _workerPublicKey Worker's signing public key, also known as "stamp"
+    * @param _workerIdentityEvidence Signature of worker's public key by worker's eth-key
+    * @param _preComputedData Additional pre-computed data for CFrag correctness verification
     **/
     function evaluateCFrag(
         bytes memory _capsuleBytes,
         bytes memory _cFragBytes,
         bytes memory _cFragSignature,
         bytes memory _taskSignature,
         bytes memory _requesterPublicKey,
-        bytes memory _stakerPublicKey,
-        bytes memory _stakerPublicKeySignature,
+        bytes memory _workerPublicKey,
+        bytes memory _workerIdentityEvidence,   // TODO: Better name (#1085)
         bytes memory _preComputedData
     )
         public
     {
-        // Check that CFrag is not evaluated yet
+        // 1. Check that CFrag is not evaluated yet
         bytes32 evaluationHash = SignatureVerifier.hash(
             abi.encodePacked(_capsuleBytes, _cFragBytes), hashAlgorithm);
         require(!evaluatedCFrags[evaluationHash], "This CFrag has already been evaluated.");
         evaluatedCFrags[evaluationHash] = true;
 
-        // Verify correctness of re-encryption
+        // 2. Verify correctness of re-encryption
         bool cFragIsCorrect = ReEncryptionValidator.validateCFrag(_capsuleBytes, _cFragBytes, _preComputedData);
         emit CFragEvaluated(evaluationHash, msg.sender, cFragIsCorrect);
-        if (cFragIsCorrect) {
-            return;
-        }
```

Comment:
This check is not removed, just postponed until we check the signatures too (see line 166). I don't think that, in case of  incorrect supporting evidence for the evaluation (e.g. bad signatures), we should (1) mark the CFrag as evaluated, and (2) emit an evaluation event, considering that the investigator already checks this off-chain before submitting (assuming she uses our nucypher code).

Reply:
Yeah, but if signatures are wrong than investigator will get just failed tx and CFrag evaluation will not be saved in storage. Checking this signatures don't change any step in algorithm if CFrag is correct: if cignatures are ok -> just check signatures, check balance of staker (require(stakerValue > 0, "Staker has no tokens");`) and do nothing - no any changes in storage state (except CFrag evaluation). if cignatures  are broken -> same checks and failed tx. Why need fail tx in that situation? 
Also from contract we can't be sure in anything outside of blockchain and can't be sure that investigator did everything properly.

Reply:
Before my changes, if CFrag was correct but signatures were wrong there was change in the state of the contract. This doesn't make sense, since it's not valid data. It may even be the case that the CFrag isn't coming from an Ursula: anyone can generate valid CFrags (e.g., using pyUmbral) with invalid signatures (since valid signatures involve participation of a real Ursula of the NuCypher network). Again, unless there's valid signatures, which imply real participation of an Ursula, we shouldn't change state.

Reply:
Can't agree that don't need to save correct CFrag when investigators sent wrong signatures: because investigator already paid and data are still valid  for contract state. But I don't want be a blocker with that so I'm ok with your point